x-pack/elastic-agent/pkg/composable/providers/kubernetes/pod.go

```diff
@@ -0,0 +1,230 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package kubernetes
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/elastic/beats/v7/libbeat/common"
+	"github.com/elastic/beats/v7/libbeat/common/safemapstr"
+
+	k8s "k8s.io/client-go/kubernetes"
+
+	"github.com/elastic/beats/v7/libbeat/common/kubernetes"
+	"github.com/elastic/beats/v7/libbeat/logp"
+	"github.com/elastic/beats/v7/x-pack/elastic-agent/pkg/agent/errors"
+	"github.com/elastic/beats/v7/x-pack/elastic-agent/pkg/composable"
+)
+
+type pod struct {
+	logger         *logp.Logger
+	cleanupTimeout time.Duration
+	comm           composable.DynamicProviderComm
+	scope          string
+}
+
+type providerData struct {
+	uid        string
+	mapping    map[string]interface{}
+	processors []map[string]interface{}
+}
+
+// NewPodWatcher creates a watcher that can discover and process pod objects
+func NewPodWatcher(
+	comm composable.DynamicProviderComm,
+	cfg *ResourceConfig,
+	logger *logp.Logger,
+	client k8s.Interface,
+	scope string) (kubernetes.Watcher, error) {
+	watcher, err := kubernetes.NewWatcher(client, &kubernetes.Pod{}, kubernetes.WatchOptions{
+		SyncTimeout:  cfg.SyncPeriod,
+		Node:         cfg.Node,
+		Namespace:    cfg.Namespace,
+		HonorReSyncs: true,
+	}, nil)
+	if err != nil {
+		return nil, errors.New(err, "couldn't create kubernetes watcher")
+	}
+	watcher.AddEventHandler(&pod{logger, cfg.CleanupTimeout, comm, scope})
+
+	return watcher, nil
+}
+
+func (p *pod) emitRunning(pod *kubernetes.Pod) {
+	data := generatePodData(pod)
+	data.mapping["scope"] = p.scope
+	// Emit the pod
+	// We emit Pod + containers to ensure that configs matching Pod only
+	// get Pod metadata (not specific to any container)
+	p.comm.AddOrUpdate(data.uid, PodPriority, data.mapping, data.processors)
+
+	// Emit all containers in the pod
+	p.emitContainers(pod, pod.Spec.Containers, pod.Status.ContainerStatuses)
+
+	// TODO deal with init containers stopping after initialization
+	p.emitContainers(pod, pod.Spec.InitContainers, pod.Status.InitContainerStatuses)
+}
+
+func (p *pod) emitContainers(pod *kubernetes.Pod, containers []kubernetes.Container, containerstatuses []kubernetes.PodContainerStatus) {
+
+	providerDataChan := make(chan providerData)
+	done := make(chan bool, 1)
+	go generateContainerData(pod, containers, containerstatuses, providerDataChan, done)
```

Comment:
Why using a channel for this? What would you think about emitting directly from the function?

Reply:
Channel usage helps to isolate the generator function so as to be possible to be tested with unit tests, following a yield-like approach.

Reply:
understood, you could also build a mocked emitter passed as a parameter to retrieve the results in the tests, right?

Reply:
Ah, good idea. Something like https://github.com/elastic/beats/pull/26801/commits/6f393782f44c3ec983b88f055d77a27159fa2551?